src/databases/WData/avtwdataFileFormat.C

```diff
@@ -0,0 +1,861 @@
+        }                                                                          \
+    }
+
+WDataVariable::WDataVariable(wdata_metadata *wdmd, int varid, int precdowngrade)
+{
+    md = wdmd;
+    vid = varid;
+    loadedcycle = -1;
+    data = NULL;
+    d2f = precdowngrade;
+}
+
+int WDataVariable::loadCycle(int cycleid)
+{
+    int ierr = 0;
+
+    if (loadedcycle != cycleid)
+    {
+        wdata_operation(wdata_read_cycle(md, md->var[vid].name, cycleid, data));
+        
+        if (ierr == 0 && d2f==1) // downgrade precision
+        {
+            int bs = (int)(wdata_get_blocksize(md, &md->var[vid]) / sizeof(double));
+            double * dd = (double*)data;
+            float * df = (float*)data;
+            for(int i=0; i<bs; i++) df[i]=(float)dd[i];
+        }
+    }
+
+    // update loadedcycle
+    if (ierr == 0)
+        loadedcycle = cycleid;
+    else
+        loadedcycle = -1;
+
+    return ierr;
+}
+
+// =======================================================================================
+// ============================= WDataRealVariable =======================================
+// =======================================================================================
+WDataRealVariable::WDataRealVariable(wdata_metadata *wdmd, int varid, int precdowngrade) : WDataVariable(wdmd, varid, precdowngrade)
+{
+    // allocate memory for data
+    int bs = wdata_get_blocklength(md);
+    data = new double[bs];
+
+    dataR = (float *)data; // for easier algebra
+
+    // create list of varaibles
+    varname.push_back(md->var[vid].name);
+    varunit.push_back(md->var[vid].unit);
+
+    // check links
+    for (int i = 0; i < md->nlink; i++)
+        if (strcmp(md->link[i].linkto, md->var[vid].name) == 0)
+        {
+            varname.push_back(md->link[i].name);
+            varunit.push_back(md->var[vid].unit);
+        }
+}
+
+bool WDataRealVariable::getVariable(const char *_varname, int cycleid, float *data_for_visit)
+{
+    // check it this varaible can generate data for varname
+    int varid = -1;
+
+    for (int i = 0; i < varname.size(); i++)
+        if (strcmp(_varname, varname[i].c_str()) == 0)
+            varid = i;
+
+    if (varid == -1)
+        return false;
+
+    // load data
+    loadCycle(cycleid);
+
+    int ix, iy, iz;
+    int ixyz1 = 0, ixyz2 = 0;
+    for (ix = 0; ix < md->nx; ix++)
+        for (iy = 0; iy < md->ny; iy++)
+            for (iz = 0; iz < md->nz; iz++)
+            {
+                ixyz2 = ix + md->nx * iy + md->nx * md->ny * iz;
+                data_for_visit[ixyz2] = (float)dataR[ixyz1];
+                //             debug4<<"[XXXX]"<<ix<<" "<<iy<<" "<<iz<<" "<<ixyz1<<" "<<ixyz2<<" "<<data[ixyz1]<<endl;
+                ixyz1++;
+            }
+
+    return true;
+}
+
+// =======================================================================================
+// ============================= WDataComplexVariable =======================================
```

Comment:
Style nitpick: There are 3 extra "=" at the end of this line.

Reply:
Done